### PR TITLE
Mepts | 1043 List Active Patients on ART without TB Screening

### PR DESCRIPTION
**Pull request by eliazmuchave**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+
+@Component
+public class ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries {
+
+  @Autowired private TxCurrCohortQueries txCurrCohortQueries;
+
+  @Autowired private TXTBCohortQueries txtbCohortQueries;
+
+  @Autowired private CommonQueries commonQueries;
+
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java:28]: Can all those members be autowired once through the constructor?

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+   *
+   * <p>Number of Patients Currently on ART without TB screening and at least one Clinical
+   * Consultation in last 6 months
+   *
+   * </blockquote>
+   *
+   * @return {@link CohortDefinition}
+   */
+  public CohortDefinition
+      getPatientsCurrentlyOnArtWithoutTbScreeningAndWithoutClinicalConsultationInLast6Months() {
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java:125]: Looks like this method is NOT referenced anywhere

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+   * </blockquote>
+   *
+   * @return {@link DataDefinition}
+   */
+  public DataDefinition getDispensationTypeOnClinicalAndPediatricEncounter() {
+
+    SqlPatientDataDefinition sqlPatientDataDefinition = new SqlPatientDataDefinition();
+    sqlPatientDataDefinition.setName("Dispensation Type on Encounter 6");
+    sqlPatientDataDefinition.addParameter(new Parameter("location", "Location", Location.class));
+    sqlPatientDataDefinition.addParameter(new Parameter("endDate", "End Date", Location.class));
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: Date taking Location object here is NOT accurate

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+   * </blockquote>
+   *
+   * @return {@link DataDefinition}
+   */
+  public DataDefinition getDispensationTypeOnFila() {
+
+    SqlPatientDataDefinition sqlPatientDataDefinition = new SqlPatientDataDefinition();
+    sqlPatientDataDefinition.setName("Dispensation Type on FILA ");
+    sqlPatientDataDefinition.addParameter(new Parameter("location", "Location", Location.class));
+    sqlPatientDataDefinition.addParameter(new Parameter("endDate", "End Date", Location.class));
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: same here

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+   * </blockquote>
+   *
+   * @return {@link DataDefinition}
+   */
+  public DataDefinition getMostRecentDrugPickupDateOnRecepcaoLevantouArv() {
+
+    SqlPatientDataDefinition sqlPatientDataDefinition = new SqlPatientDataDefinition();
+    sqlPatientDataDefinition.setName("Most Recent Drug Pick-Up on Recepção Levantou ARVs ");
+    sqlPatientDataDefinition.addParameter(new Parameter("location", "Location", Location.class));
+    sqlPatientDataDefinition.addParameter(new Parameter("endDate", "End Date", Location.class));
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: Same here

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+   * </blockquote>
+   *
+   * @return {@link DataDefinition}
+   */
+  public DataDefinition getNextScheduledDrugPickupDateOnRecepcaoLevantouArv() {
+
+    SqlPatientDataDefinition sqlPatientDataDefinition = new SqlPatientDataDefinition();
+    sqlPatientDataDefinition.setName("Most Recent Drug Pick-Up on Recepção Levantou ARVs ");
+    sqlPatientDataDefinition.addParameter(new Parameter("location", "Location", Location.class));
+    sqlPatientDataDefinition.addParameter(new Parameter("endDate", "End Date", Location.class));
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: Map this to date object

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+
+    Map<String, Integer> valuesMap = new HashMap<>();
+    valuesMap.put("23866", hivMetadata.getArtDatePickupMasterCard().getConceptId());
+    valuesMap.put("52", hivMetadata.getMasterCardDrugPickupEncounterType().getEncounterTypeId());
+
+    String query =
+        "SELECT p.patient_id, "
+            + "       MAX(DATE_ADD(o.value_datetime, INTERVAL 30 MONTH)) encounter_date "
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: Looks like this interval range need to be in days, NOT months

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+    valuesMap.put("165174", hivMetadata.getLastRecordOfDispensingModeConcept().getConceptId());
+
+    String query =
+        "SELECT p.patient_id, Max(e.encounter_datetime) consultation_date "
+            + "FROM   patient p "
+            + "       INNER JOIN encounter e ON e.patient_id = p.patient_id "
+            + "       INNER JOIN obs o ON o.encounter_id = e.encounter_id "
+            + "WHERE  e.encounter_type = ${6} "
+            + "       AND e.location_id = :location "
+            + "       AND e.encounter_datetime <= CURRENT_DATE() "
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java:421]: Do you want to compare against current time and date of you will need to have end of reporting period?

Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java:421]: Maybe report generation date could mean the same thing, current date when the report is being generated

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+
+    return sqlCohortDefinition;
+  }
+
+  private void addParameters(CohortDefinition cd) {
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+  }
+
+  private void addSqlCohortDefinitionParameters(SqlCohortDefinition sqlCohortDefinition) {
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: Not referenced anywhere

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListOfPatientsCurrentlyOnArtWithoutTbScreeningDataset.java [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+  @Autowired private TPTInitiationDataDefinitionQueries tptInitiationDataDefinitionQueries;
+
+  @Autowired
+  private ListOfPatientsEligibleForVLDataDefinitionQueries
+      listOfPatientsEligibleForVLDataDefinitionQueries;
+
+  @Autowired private TxtbDenominatorQueries txtbDenominatorQueries;
+
+  @Autowired private HivMetadata hivMetadata;
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListOfPatientsCurrentlyOnArtWithoutTbScreeningDataset.java]: Collapse all those Autowired calls into one encapsulated in a constructor

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListOfPatientsCurrentlyOnArtWithoutTbScreeningDataset.java [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+            .getPatientsCurrentlyOnArtWithoutTbScreening(),
+        "endDate=${endDate},location=${location}");
+
+    PatientIdentifierType identifierType =
+        Context.getPatientService()
+            .getPatientIdentifierTypeByUuid("e2b966d0-1d5f-11e0-b929-000c29ad1d07");
+
+    DataConverter identifierFormatter = new ObjectFormatter("{identifier}");
+
+    DataDefinition identifierDef =
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListOfPatientsCurrentlyOnArtWithoutTbScreeningDataset.java]: Not used in the dataset, was it supposed to be the NID?

Comment (ningosi) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListOfPatientsCurrentlyOnArtWithoutTbScreeningDataset.java]: if NOT going to be used, please consider removing

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/reports/SetupListOfPatientsCurrentlyOnArtWithoutTbScreening.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+import org.openmrs.module.reporting.ReportingException;
+import org.openmrs.module.reporting.evaluation.parameter.Mapped;
+import org.openmrs.module.reporting.evaluation.parameter.Parameter;
+import org.openmrs.module.reporting.report.ReportDesign;
+import org.openmrs.module.reporting.report.definition.ReportDefinition;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class SetupListOfPatientsCurrentlyOnArtWithoutTbScreening extends EptsDataExportManager {
```
Comment (ningosi) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/reports/SetupListOfPatientsCurrentlyOnArtWithoutTbScreening.java:23]: No any base cohort is needed for this report to limit number of patients returned?

**Review by ningosi — COMMENTED** [review]
@eliazmuchave left some comments, please have a look

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java (hunk 1 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java:28]: Ok

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java (hunk 2 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java:125]: Fixed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java (hunk 3 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: Fixed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java (hunk 4 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: Fixed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java (hunk 5 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java (hunk 6 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java (hunk 7 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: Yes. Fixed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java (hunk 8 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java:421]: Implemented cording to the requirements

The system will identify patient’s most recent records of MDCs (DSD) registered in Ficha Clínica by report generation date and will show the registered MDC1, MDC2, MDC3, MDC4 and MDC5 if marked as “Inicio” or “Continua” as follows:

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java (hunk 9 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ListOfPatientsCurrentlyOnArtWithoutTbScreeningCohortQueries.java]: fixed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListOfPatientsCurrentlyOnArtWithoutTbScreeningDataset.java (hunk 10 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListOfPatientsCurrentlyOnArtWithoutTbScreeningDataset.java]: done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListOfPatientsCurrentlyOnArtWithoutTbScreeningDataset.java (hunk 11 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListOfPatientsCurrentlyOnArtWithoutTbScreeningDataset.java]: Removed

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/reports/SetupListOfPatientsCurrentlyOnArtWithoutTbScreening.java (hunk 12 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/reports/SetupListOfPatientsCurrentlyOnArtWithoutTbScreening.java:23]: For this report all patients are picked from tx-curr and tx-curr already implemented the basic base cohort. I dont know if in this scenario there is a need to include the basic base cohort in the setup

**Review by ningosi — APPROVED** [review]
Review done, will be validated on the runned results

**Pull request merged**